### DISCUSSION TITLE
[Q&A] LangChain or Vercel AI SDK

### DISCUSSION THREAD

**Discussion by zandko** [lobehub/lobehub#277] [Development Question] [answered]
### 🧐 问题描述 | Proposed Solution

我注意到在先前的代码中使用了langchain库，但现在改为使用openai库。langchain似乎更方便，但是为什么还是选择了openai呢？

**Comment by arvinxx** [comment, marked as answer]
几个点：

1. 早期版本的 LangChain 和 Edge Runtime 兼容性不是非常好，关于 stream 的实现需要自己手撸，而且相应的包体积也大很多。
2. 后来看了  [vercel ai sdk](https://sdk.vercel.ai/docs/guides/providers/openai#create-a-route-handler) 的文档，换成了 openai 原生包 + `ai` 的 StreamResponse 方法，代码量少了不少，心智负担也减少了，详见：https://github.com/lobehub/lobe-chat/commit/170567a87f221a6dfbf5051fbc2cbf8b10803fd4#diff-712dc70c79419ec0fa9c068859d6d9874790e5d8f7b6b9ce8bb992410d4a22cc；
3. 目前 LobeChat 还没有用到 vector db、 embedding 、 chain 等属于 langchain 擅长的领域的高阶能力，所以暂时也不需要引入 LangChain；
4.  LangChain 的 ts 版本，其实有挺多细节问题的，比如：https://github.com/langchain-ai/langchainjs/issues/1082 。上半年用 ts 版本 prompt template ，就遇到了这个问题，结果半年过去了都没修就很离谱… 如果你自己看看它的实现，会发现很裸，可维护性也挺有问题的。所以我个人是有点怀疑 LangChain 作者在 ts 包上的专业性来着…
